app/src/androidTest/java/com/example/antitheft/structure/GPSCommandHandlerTest.java

```diff
@@ -1,6 +1,106 @@
 package com.example.antitheft.structure;
 
+import com.eis.smslibrary.SMSMessage;
+import com.eis.smslibrary.SMSPeer;
+
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * @author Francesco Bau'
+ * @since 29/02/2020
+ */
 public class GPSCommandHandlerTest {
 
+    private SMSMessage smsMessage;
+    private SMSPeer smsPeer;
+
+    private GPSCommandHandler handler;
+
+    static final String LOCATION = "";
+    // WARNING: insert a VALIID telephoneNumber, because it sends an SMS to that telephone number.
+    static final String VALID_TELEPHONE_NUMBER = "+393515186755";
+    static final String TEXT_MESSAGE = "";
+
+
+    /**
+     * Initialization of the fields, before testing.
+     *
+     * @see com.eis.smslibrary.SMSPeer
+     * @see com.eis.smslibrary.SMSMessage
+     */
+    @Before
+    public void init() {
+        smsPeer = new SMSPeer(VALID_TELEPHONE_NUMBER);
+        smsMessage = new SMSMessage(smsPeer, TEXT_MESSAGE);
+        handler = new GPSCommandHandler();
+    }
+
+    /**
+     * Test what happens if there's an attempt to send a null {@link SMSMessage} .
+     * <p>
+     * {@link NullPointerException} is expected to be thrown, because of the @NonNull annotation.
+     *
+     * @see NullPointerException
+     * @see com.eis.smslibrary.SMSMessage
+     * @see GPSCommandHandler#sendMessage(SMSMessage)
+     */
+    @Test(expected = NullPointerException.class)
+    public void sendNullMessage() {
+        handler.sendMessage(null);
+    }
+
+    /**
+     * Test what happens if a null {@link SMSMessage} is received in method
+     * {@link GPSCommandHandler#onCommandReceived(SMSMessage)}.
+     * <p>
+     * {@link NullPointerException} is expected to be thrown, because of the @NonNull annotation.
+     *
+     * @see NullPointerException
+     * @see com.eis.smslibrary.SMSMessage
+     * @see GPSCommandHandler#onCommandReceived(SMSMessage)
+     */
+    @Test(expected = NullPointerException.class)
+    public void onNullCommandReceived() {
+        handler.onCommandReceived(null);
+    }
+
+    /**
+     * Test what happens if there's an attempt to call {@link GPSCommandHandler#sendLocation(SMSPeer, String)}
+     * with a null {@link SMSPeer}.
+     * <p>
+     * {@link NullPointerException} is expected to be thrown, because of the @NonNull annotation.
+     *
+     * @see NullPointerException
+     * @see com.eis.smslibrary.SMSPeer
+     * @see GPSCommandHandler#sendLocation(SMSPeer, String)
+     */
+    @Test(expected = NullPointerException.class)
+    public void sendLocationToNullPeer() {
+        handler.sendLocation(null, LOCATION);
+    }
+
+    /**
+     * Test what happens when a certain {@link SMSMessage}, supposed to be a command, is received by
+     * method {@link GPSCommandHandler#onCommandReceived(SMSMessage)}.
+     *
+     * @see com.eis.smslibrary.SMSMessage
+     * @see GPSCommandHandler#onCommandReceived(SMSMessage)
+     */
+    @Test
+    public void onCommandReceived() {
```

Comment:
This method throws java.lang.NullPointerException for a null reference problem on the MainActivity.

Reply:
Oops, my bad!
Thank you for spotting that...

Reply:
> This method throws java.lang.NullPointerException for a null reference problem on the MainActivity.

I was wondering... Am I supposed to add a `@ throws` declaration in method `getCurrentLocation()` as well?

Reply:
Wait... That's inside a Test class... I don't know if I'm supposed to do that, it should just fail...

Reply:
It's not supposed to throw `NullPointerException`, but it does, because the method `getCurrentLocation` in `MainActivity` is not well-implemented. That method should be inside a Service, and I'm gonna do that in `develop` branch.
It's thanks to that test that I was able to know why does it not work at `onDestroy()` . So I can't remove it or "mask it" into something I expect. I'm gonna clarify it as a Known Issue, though... My apologies for not being clear enough.